### PR TITLE
fix: ip prefix bucket reduction

### PR DESCRIPTION
**Pull request by salvatore-campagna**
When reducing buckets coming from different shards we need to
aggregate the doc count from all the buckets with the same key.
Here, when calling the 'reduceBucket' method all the 'currentBuckets'
share the same key. As a result, we just need to create a new
bucket with the correct doc count value.

This bug does not affect aggregations in a cluster with just one
shard because in that case 'currentBuckets' is just a list with
a single bucket with a specific key coming from the unique shard.

This is a fix for bug #83321.

### CODE REVIEW

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/450_ip_prefix.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ setup:
         body:
           settings:
             number_of_replicas: 0
+            number_of_shards: 3
```
Comment (salvatore-campagna) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/450_ip_prefix.yml:8]: @nik9000 do you think it is ok to force here a specific number of shards or should we rely on "randomness" not specifying the value? As it is with this PR we might not spot issues for the "one-shard" scenario.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/450_ip_prefix.yml [hunk 2]
```diff
@@ -22,25 +23,25 @@ setup:
         refresh: true
         body:
           - { "index": { } }
-          - { "ipv4": "192.168.1.10", "ipv6": "2001:db8:a4f8:112a:6001:0:12:7f10", "value": 10, ip: "192.168.1.10" }
+          - { "ipv4": "192.168.1.10", "ipv6": "2001:db8:a4f8:112a:6001:0:12:7f10", "value": 10, "ip": "192.168.1.10" }
```
Comment (salvatore-campagna) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/450_ip_prefix.yml:26]: I added double quotes here around the field name just for consistency.

File: docs/changelog/83637.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+pr: 83637
+summary: "Fix: ip prefix bucket reduction"
+area: Aggregations
+type: bug
+issues: []
```
Comment (salvatore-campagna) [inline docs/changelog/83637.yaml:5]: Do I need to add 83321 here?

**Comment by salvatore-campagna** [comment]
I think that BWC tests are failing because of the bug affecting version `8.1.0` which is fixed by this PR only for version `8.2.0`.

File: docs/changelog/83637.yaml (hunk 3 above)
Comment (nik9000) [reply docs/changelog/83637.yaml:5]: I have no idea. I don't really understand this process totally.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/450_ip_prefix.yml (hunk 1 above)
Comment (nik9000) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/450_ip_prefix.yml:8]: I'd just rely on the randomness. You can always reproduce it with the failed hash.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/450_ip_prefix.yml (hunk 2 above)
Comment (nik9000) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.aggregation/450_ip_prefix.yml:26]: Sure

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/InternalIpPrefix.java [hunk 4]
```diff
@@ -318,15 +318,30 @@ public Bucket createBucket(InternalAggregations aggregations, Bucket prototype)
         );
     }
 
+    public Bucket createBucket(Bucket prototype, InternalAggregations aggregations, long docCount) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/InternalIpPrefix.java:321]: If this is only used in this method I think it should be private. I think other aggs have something like this be public but I don't know why. I think it was a historical accident.

**Review by nik9000 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/InternalIpPrefix.java (hunk 4 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/InternalIpPrefix.java]: Of course.

**Review by nik9000 — APPROVED** [review]

**Pull request merged**